community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/spi/RepeatableReadQueryContext.scala

```diff
@@ -113,11 +114,51 @@ final class RepeatableReadQueryContext(inner: QueryContext, locker: Locker) exte
     locker.acquireLock(relationshipOps.getByInnerId(id))
   }
 
-  private def lockAll[T <: PropertyContainer](iter: Iterator[T]): Iterator[T] = iter.map {
-    item =>
-      locker.acquireLock(item)
-      item
+
+  private def lockAllNodesAndFilter(inner: Iterator[Node])(f: Node => Boolean): Iterator[Node] = new Iterator[Node] {
+    var nextNode: Option[Node] = null
```

Comment:
I don't think this is a bad solution. There are explicitly checks for the three different types of values this field can have - Some/None and null. Option is still being used as an Option, it's just that un-initialised is flagged with null.